### PR TITLE
Walkthrough Tutorial

### LINKED ISSUE #772: RFC: Instructions and examples
### Conclusion
Implement a walkthrough tutorial.

### More to discuss
- Do we want to use Intro.js or React Joyride?
- What features do we want to highlight in the tutorial and in what order? We don't need to fully enumerate yet, just say what's important.

### Problem
I was talking to my HCI group, and one person said, I think validly, that it's not obvious what to do when you open the site, and that's intimidating. There are a lot of features, but nothing telling people what to do/how to do them. 

### Possible approaches
- Interactive tutorial. Point to things and let them click around.
- One-time overlay pointing to the basic features.
- Video tutorial(s)
- Something else?

### Related Issues
#140

### PR DESCRIPTION
**Pull request by MinhxNguyen7**
## Summary
Semi-interactive tutorial for user onboarding. Justification is given in #772.

Tutorial library: [reactour](https://github.com/elrumordelaluz/reactour). We initially considered intro.js-react and ReactJoyride, but neither of them functioned properly with our app during testing.

### Tutorial Steps (In Progress)
1) Search bar.
    - Automatically add a class to calendar after proceeding?
    - Import button?
2) Class schedule.
3) Finals button.
4) Added courses.
5) Map.
6) Save button.
7) Load button.
8) Export to .ics. Be verbose. Get inspiration from the tooltip in the download button

## Request For Comment
- What should be added or removed? Both are important. stupidly long onboarding tutorials are stupid.
- Does the ordering need to change?
- When should the tutorial be showed or not be showed?
    - If the user dismisses it, don't show it next time.
    - What if the user leaves the page, should it show next time? I'm leaning towards yes.

## TODO
- [x] Choose and test tutorial library.
- [x] Enumerate the tutorial steps.
- [x] Implement the steps.
- [x] Make sure it functions and looks decent.
    - [x] Style tooltip.
    - [x] Dark mode.
- [x] Disable tutorial on subsequent visits.
    - [x] Including if someone has schedule in localstorage.
- [x] Disable tutorial when the patch notes are up.
- [x] MUI FAB button with low opacity on subsequent visits.
- [x] Click buttons for users if they skip through.

## Test Plan
TBA

## Issues

Closes #772

### CODE REVIEW

**Comment by EricPedley** [comment]
Pressing esc doesn't work on firefox until you click on the page

**Comment by MinhxNguyen7** [comment]
> Pressing esc doesn't work on firefox until you click on the page

It works on Chrome/Chromium and Safari. We'll just call it a Firefox L for now.

File: apps/antalmanac/src/components/RightPane/CoursePane/CourseRenderPane.tsx [hunk 1]
```diff
@@ -21,10 +21,13 @@ import { openSnackbar } from '$actions/AppStoreActions';
...
-        accumulator[section.sectionCode] = section.color;
-        return accumulator;
-    }, {} as { [key: string]: string });
+    const courseColors = AppStore.schedule.getCurrentCourses().reduce(
+        (accumulator, { section }) => {
+            accumulator[section.sectionCode] = section.color;
+            return accumulator;
+        },
+        {} as { [key: string]: string }
+    );
```
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/CoursePane/CourseRenderPane.tsx:31]: chrome-capture-2024-0-16

Step 2 breaks if course render pane is shown, instead of the search form

File: apps/antalmanac/src/lib/tourExampleGeneration.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            bldg,
+            days,
+            startTime,
+            endTime,
+            timeIsTBA,
+        },
+    ];
+}
+
+export function sampleFinalExamFactory({
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/tourExampleGeneration.ts:153]: Total nit that I'm sure you'll get around to -- but just to be overly thorough, the sample courses should have their data fully filled out.

Screenshot 2024-01-16 at 12 47 52 PM

File: apps/antalmanac/src/lib/TutorialHelpers.tsx [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        addedCoursePane: {
+            selector: '#course-pane-box',
+            content: (
+                <>
+                    <b>Select</b> the added courses tab for a list of your courses and details
+                </>
+            ),
+            action: () => setTab(1),
+            mutationObservables: ['#course-pane-box'],
+        },
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/TutorialHelpers.tsx:162]: Because the text here is the same, I'm wondering if we can have `addedCoursePane` highlight/mention some of the features on the added pane.

File: apps/antalmanac/src/lib/TutorialHelpers.tsx [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+                selector: '#map-tab',
+                eventType: 'click',
+            }),
+        },
+        mapPane: {
+            selector: '#map-pane',
+            content: <>Select the map tab to see where your classes are.</>,
+            action: () => setTab(2),
+            mutationObservables: ['#map-pane'],
+        },
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/TutorialHelpers.tsx]: Same point as above -- perhaps map pane can refer to the markers and bring up the routing?

File: apps/antalmanac/src/lib/tourExampleGeneration.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            bldg,
+        };
+
+    return {
+        examStatus,
+        dayOfWeek: dayOfWeek ?? randomWeekday(),
+        month,
+        day,
+        startTime: startTime ?? randomClasstime(),
+        endTime: endTime ?? randomClasstime(),
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/tourExampleGeneration.ts]: Not sure if this should be random; Is there an advantage over have a specific fallback time?

File: apps/antalmanac/src/lib/TutorialHelpers.tsx [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                fontFamily: 'monospace',
+                padding: '0.05rem 0.3rem',
+            }}
+        >
+            {props.children}
+        </div>
+    );
+}
+
+export function namedStepsFactory(goToStep: (step: number) => void): Record<TourStepName, StepType> {
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/TutorialHelpers.tsx:56]: Can we cleanup the added tutorial courses afterwards? Probably a rounding error overall, but I see no reason to keep them since it's not difficult to just start adding courses.

File: apps/antalmanac/src/lib/TutorialHelpers.tsx [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+            /** If the user presses/pressed the finals button, move to the next step */
+            action: tourActionFactory(
+                () => {
+                    goToNamedStep(TourStepName.showFinals);
+                },
+                // Only move to the next step if the button is toggled from off to on, not vice versa.
+                { selector: '#finals-button', eventType: 'click' }
+            ),
+        },
+        showFinals: {
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/TutorialHelpers.tsx:124]: Can we unclick the button after this step? Currently the tutorial won't make sense if the button is already clicked. 

In the same vein, we should set the finals button to unclicked before entering the `calendar` step, so that the appropriate calendar is shown.

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Fantastic stuff, great PR! Very impressed by the tutorial -- it felt concise yet informative, great balance.

My only feedback is on some nitpicks and some suggestions regarding flow. Will do more QA testing and such as we go through the review process.

As for the RFC:
1. Don't feel a need to add or remove a step, feels good right now
2. Order is good
3. Agree that if dismissed, don't show next time (they can always click the FAB). If tutorial remains uninteracted with, and page is closed, then show next time. TL;DR I'm in agreement with your thought process.

File: apps/antalmanac/src/lib/tourExampleGeneration.ts (hunk 2 above)
Comment (MinhxNguyen7) [reply apps/antalmanac/src/lib/tourExampleGeneration.ts:108]: I wasn't going to get around to that. I didn't even notice that I'd written this strange thing.

File: apps/antalmanac/src/lib/tourExampleGeneration.ts (hunk 5 above)
Comment (MinhxNguyen7) [reply apps/antalmanac/src/lib/tourExampleGeneration.ts]: There isn't. I just didn't want to spell them out.

File: apps/antalmanac/src/lib/TutorialHelpers.tsx (hunk 3 above)
Comment (MinhxNguyen7) [reply apps/antalmanac/src/lib/TutorialHelpers.tsx:162]: I think it'd be better not to because the page doesn't have particularly important stuff, so we shouldn't go into it in too much detail.

File: apps/antalmanac/src/lib/TutorialHelpers.tsx (hunk 4 above)
Comment (MinhxNguyen7) [reply apps/antalmanac/src/lib/TutorialHelpers.tsx]: This, though, has a lot of features that are worth talking about.

File: apps/antalmanac/src/components/Map/Map.tsx [hunk 8]
```diff
@@ -181,6 +184,20 @@ export default function CourseMap() {
...
+        };
+
+        AppStore.on('customEventsChange', updateCustomEventMarkers);
+        AppStore.on('currentScheduleIndexChange', updateCustomEventMarkers);
+
+        return () => {
+            AppStore.removeListener('customEventsChange', updateCustomEventMarkers);
+            AppStore.removeListener('currentScheduleIndexChange', updateCustomEventMarkers);
+        };
+    }, []);
```
Comment (KevinWu098) [inline apps/antalmanac/src/components/Map/Map.tsx]: Code snuck in from the merge, duplicate from 173-183
```suggestion
```

**Comment by KevinWu098** [comment]
Changes look good at first pass, will delve deeper post IH

File: apps/antalmanac/src/lib/TutorialHelpers.tsx [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        importButton: {
+            selector: '#import-button',
+            content: 'Quickly add your classes from WebReg or Zotcourse!',
+        },
+        calendar: {
+            selector: '.rbc-time-view', // Calendar.
+            content: 'See the classes in your schedule!',
+            action: addSampleClasses,
+        },
+        finalsButton: {
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/TutorialHelpers.tsx]: If the Finals button is already clicked when entering the tutorial, these two steps don't make sense.

File: apps/antalmanac/src/lib/tourExampleGeneration.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            endTime,
+            timeIsTBA,
+        },
+    ];
+}
+
+export function sampleFinalExamFactory({
+    examStatus = 'SCHEDULED_FINAL',
+    dayOfWeek,
+    month = 12,
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/tourExampleGeneration.ts]:
```suggestion
    month = 11,
```
0 indexed months

File: apps/antalmanac/src/components/Tutorial.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    ]);
+
+    const restartTour = useMemo(
+        () => () => {
+            displaySearch();
+            disableManualSearch();
+            setCurrentStep(0);
+            setIsOpen(true);
+        },
+        [setCurrentStep, setIsOpen]
```
Comment (KevinWu098) [inline apps/antalmanac/src/components/Tutorial.tsx]:
```suggestion
        [disableManualSearch, displaySearch, setCurrentStep, setIsOpen]
```
exhaustive-deps

File: apps/antalmanac/src/lib/tourExampleGeneration.ts [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    day = 12,
+    startTime,
+    endTime,
+    bldg = [],
```
Comment (KevinWu098) [inline apps/antalmanac/src/lib/tourExampleGeneration.ts]:
```suggestion
    bldg = [DBH],
```
Have some building for the CourseCalendarEvent popup

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Just a few other things I caught, otherwise looks good to merge.

File: apps/antalmanac/src/components/Map/Map.tsx [hunk 13]
```diff
@@ -171,16 +171,16 @@ export default function CourseMap() {
...
-        AppStore.on('addedCoursesChange', updateCalendarEvents);
-        AppStore.on('currentScheduleIndexChange', updateCalendarEvents);
+        AppStore.on('customEventsChange', updateCustomEventMarkers);
+        AppStore.on('currentScheduleIndexChange', updateCustomEventMarkers);
 
         return () => {
-            AppStore.removeListener('addedCoursesChange', updateCalendarEvents);
-            AppStore.removeListener('currentScheduleIndexChange', updateCalendarEvents);
+            AppStore.removeListener('customEventsChange', updateCustomEventMarkers);
+            AppStore.removeListener('currentScheduleIndexChange', updateCustomEventMarkers);
```
Comment (KevinWu098) [inline apps/antalmanac/src/components/Map/Map.tsx]: Unless I'm mistaken, I believe these lines should not have been changed. Mixup from duplicate map marker code, which admittedly we maybe shouldn't be handling here in an unrelated PR but uh... why not 🤷‍♂️ 

TL;DR revert these lines of code and we gouda

(from main):
```tsx
    useEffect(() => {
        const updateCalendarEvents = () => {
            setCalendarEvents(AppStore.getCourseEventsInCalendar());
        };

        AppStore.on('addedCoursesChange', updateCalendarEvents);
        AppStore.on('currentScheduleIndexChange', updateCalendarEvents);

        return () => {
            AppStore.removeListener('addedCoursesChange', updateCalendarEvents);
            AppStore.removeListener('currentScheduleIndexChange', updateCalendarEvents);
        };
    }, []);
```

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Banger PR, excited to merge; Just one thing to fix and we're good to go!

**Review by KevinWu098 — APPROVED** [review]
LGTM 🚀
